v3/src/liftoff/core.ts

```diff
@@ -0,0 +1,101 @@
+import { trace, Bond } from './pattern'
+import { Ref, isRef, DEFAULT_BOND, getLabel } from './ref'
+import { throws } from './errors'
+import { getLocation } from './loc'
+
+export type Plan = () => void
+
+export class Core {
+  private defs: Map<Ref<any>, Set<Bond>> = new Map
+
+  constructor(public readonly plan: Plan) {
+    this.link(plan)
+  }
+
+  link(plan: Plan) {
+    const {defs} = this
+    for (const delta of trace(plan)) {
+      const {mut, bond} = delta
+      if (mut === 'add') {
+        if (bond.type === 'def') {
+          const {state: {ref}} = bond
+          if (!defs.has(ref))
+            defs.set(ref, new Set([bond]))
+          else
+            defs.get(ref)!.add(bond)
+        }
+      }
+      // TODO (queerviolet): Handle change and delete deltas
+    }
+  }
+
+  states: Map<Ref<any>, any> = new Map
+
+  /**
+   * Resolve all bonds that define this ref.
+   *
+   * @param ref {Ref<T>} the ref
+   * @returns Set<Bond>
+   */
+  private resolveDefs(ref: Ref<any>): Set<Bond> {
+    return this.defs.get(ref) || (
+      typeof ref[DEFAULT_BOND] !== 'undefined' ? new Set([ref[DEFAULT_BOND]]) : NO_BONDS
+    )
+  }
+
+  /**
+   * Given a ref, return an array of all defined values for that ref.
+   *
+   * @param ref {Ref<T>} the ref to read
+   * @returns Promise<T[]>
+   */
+  async once<T>(ref: Ref<T>): Promise<T[]> {
```

Comment:
Yeah, I'm refactoring to eliminate the `once` / `only` distinction now. It's going well, and I think it'll simplify a lot.

What I'm working towards:

- `Ref<T>` becomes instead: `Ref<Output=any, Input=Output>`
  - `Input` is what you can call a `Ref` with to define it
  - `Output` is what you get back when you read it
- _Quick helpers for the next few examples:_

  ```ts
  type InputOf<R extends Ref> = R extends Ref<any, infer I> ? I : never
  type OutputOf<R extends Ref> = R extends Ref<infer O, any> ? O : never
  ```

- The core gets a single `get` method:
  ```ts
  get<R extends Ref>(ref: R): OutputOf<R>
  ```

  - This method is intentionally not async
  - The behavior of `get` on scalars mostly matches the current behavior of `only`—return the single def, fail if defined multiple times. One change, I think: we'll return `undefined` rather than throwing if it's undefined.
- If you want all the values of a ref which has been defined multiple times, you can call `get(many(ref))`, where `many` is:

    ```ts
    function many<R extends Ref>(ref: R): Ref<OutputOf<R>[], InputOf<R>>
    ```
  
  - If you want a ref that always behaves this way, you can call `many` on it when you define it!

    ```ts
    const Schema = many(obj <GraphQLSchema> ())
    ```
    - Because `many` doesn't change the ref's input type, you still define schemas with `Schema({ typeDefs: ... })`
    - This (and some forward-looking concerns, like representing open pipelines with `map` and `reduce`) are the motivation between separating out the input and output types of a ref.

- All these `get` calls are synchronous, and return the defined values according to the current state. When we introduce asynchrony and reactivity, we can introduce new ops like e.g.:

  ```ts
  function once<R extends Ref>(ref: R): Ref<Promise<OutputOf<R>>, InputOf<R>>
  function asyncIterable<R extends Ref>(ref: R): Ref<AsyncIterable<OutputOf<R>>, InputOf<R>>
  ```

The best bit here is that most of the implementation stays the same, and a good chunk of it goes away.

Would love to hear any thoughts you have on this direction!